Improve compaction task status updates

davisp opened the pull request:
Previous the emsort related operations did not update the compaction
task status. For large databases this leads to some very long waits
while the compaction task stays at 100%. This change adds progress
reports to the steps for sorting and copying document ids back into the
database file.

davisp committed:
Improve compaction task status updates

Previous the emsort related operations did not update the compaction
task status. For large databases this leads to some very long waits
while the compaction task stays at 100%. This change adds progress
reports to the steps for sorting and copying document ids back into the
database file.

iilyak commented on src/couch_db_updater.erl:
```diff
@@ -1323,12 +1326,82 @@ bind_id_tree(Db, Fd, State) ->
+
+emsort_cb(_Ems, {merge, chain}, {init, Copied, Nodes}) ->
+    {init, Copied, Nodes + 1};
+emsort_cb(_Ems, row_copy, {init, Copied, Nodes}) when Copied >= 1000 ->
```
Magic constant 1000.

davisp replied:
I assume you mean you'd rather that be a define? I just copied the same shape as merge_docids down below.

iilyak replied:
Yes I mean that it would be better to have a define. Like ?UPDATE_FREQ or ?BATCH_SIZE or something.

iilyak commented on src/couch_db_updater.erl:
```diff
@@ -1323,12 +1326,82 @@ bind_id_tree(Db, Fd, State) ->
+    ]),
+    0;
+
+emsort_cb(_Ems, row_copy, Copied) when is_integer(Copied), Copied > 1000 ->
```
Magic constant 1000.

iilyak commented on src/couch_db_updater.erl:
```diff
@@ -1323,12 +1326,82 @@ bind_id_tree(Db, Fd, State) ->
+    {init, 0, Nodes};
+emsort_cb(_Ems, row_copy, {init, Copied, Nodes}) ->
+    {init, Copied + 1, Nodes};
+emsort_cb(Ems, {merge_start, reverse}, {init, Copied, Nodes}) ->
```
I noticed that we can also have `{merge_start, forward}` [event](https://github.com/apache/couchdb-couch/blob/1db1337301a7c897be41e13cd328270509008478/src/couch_emsort.erl#L246:L252). Should we handle it as well?

iilyak replied:
It looks like we wouldn't have a progress updates for the second pass. Is it intentional?

davisp commented:
Also I need to change my use of DocCount to be TotalChanges. I noticed while testing the optimizations I tried that its currently not correct during compaction retries.

iilyak reviewed (commented):
It looks good but I have few questions.

davisp commented:
Updated to use define's for batch sizes and fixed the changes/doc count mixup.

For the {merge_start, forward}, thats only there for completeness. It felt a bit weird to not include that event even though I don't necessarily need it for this work.

Updates during the second phase happen here:

https://github.com/apache/couchdb-couch/pull/241/files#diff-f6f654ab26b490bab95be6f502c49d89R1376

This is a bit subtle but it seemed like the best I could do without starting to modify the on-disk contents of files which I like to avoid when at all possible.

Its a bit funky but without starting to store the total number of rows in emsort the best approach I had was to guess with the input being the total number of changes processed in this run and then in the first phase of the merge sort just count how many rows we have. Once that first pass is over we can calculate the total number of rows that will be copied and then just update progress in the normal fashion. Its a bit awkward but that emsort_cb kind of switches modes once it sees that the first phase has ended and then its purely just listening for row_copy events.

davisp commented:
And I hesitate to store the total number of rows because there'd then be upgrade things to worry about and generally speaking this will just sort itself out while still giving an approximate progress update.
